src/core/Pipelining.ts

```diff
@@ -0,0 +1,181 @@
+ * problems. For testing purposes we frequently have a Pipelining of 1000 or more
+ * concurrent requests to be able to saturate the system.
+ *
+ * A Pipelining provides its own backpressure on the system. So there will not be more
+ * in flight invocations than the depth of the Pipelining. This means that the Pipelining
+ * will work fine when backpressure on the member is disabled (default). Also
+ * when it it enabled it will work fine, but keep in mind that the number of concurrent
+ * invocations in the Pipelining could be lower than the configured number of invocation
+ * of the Pipelining because the backpressure on the member is leading.
+ *
+ * The Pipelining has been marked as Beta since we need to see how the API needs to
+ * evolve. But there is no problem using it in production. We use similar techniques
+ * to achieve high performance.
+ *
+ * @param E
+ */
+export class Pipelining<E> {
+    private readonly depth: number;
+    private readonly next: () => void;
+    private loadGenerator: () => Promise<E>;
+    private rejected: boolean = false;
+    private done: boolean = false;
+    private resolve: (thenableOrResult?: (void | E[]) | PromiseLike<void | E[]>) => void;
+    private reject: (err?: any) => void;
+    private resolvingCount: number = 0;
+    private index: number = 0;
+    private results: E[] = [];
+
+    /**
+     * Creates a Pipelining with the given depth and load generator.
+     *
+     * @param depth the maximum number of concurrent calls allowed in this Pipelining.
+     * @param loadGenerator a generator-like function that returns a promise
+     *  or null if the generator is exhausted.
+     * @param storeResults a boolean value indicating that whether the results of the requests
+     *  should be stored in the Pipelining or not. If this value is set to false, it is expected
+     *  from the user to handle the results of the requests within the load generator. It is false
+     *  by default.
+     * @throws {AssertionError} if the depth is smaller than 1.
+     * @throws {AssertionError} if the load generator is null.
+     */
+    constructor(depth: number, loadGenerator: () => Promise<E>, storeResults: boolean = false) {
+        assertPositive(depth, 'depth should be positive');
+        assertNotNull(loadGenerator, 'load generator cannot be null');
+        this.depth = depth;
+        this.loadGenerator = loadGenerator;
+        if (storeResults) {
+            this.next = this.nextWithResults;
+        } else {
+            this.next = this.nextWithoutResults;
+        }
+    }
+
+    /**
+     * If the Pipelining is constructed to store the results, results will be
+     * returned with order the requests were done when all of the requests
+     * are completed.
+     *
+     * If not, it will return void when all of the requests are completed.
+     *
+     * @throws {Error} if any of the requests fails
+     * @returns array of the results or void
+     */
+    run(): Promise<void | E[]> {
```

Comment:
What happens if i do
```
let pipe = new Pipelining(...);
pipe.run();
pipe.run();
```
Looks like all internal variables are going to be messed up. We need to prevent concurrent restarts.